File: src/main/java/chat/teco/tecochat/like/chatlike/application/service/ChatLikeService.java
package chat.teco.tecochat.like.chatlike.application.service;

import static chat.teco.tecochat.chat.exception.chat.ChatExceptionType.NOT_FOUND_CHAT;

import chat.teco.tecochat.chat.domain.chat.Chat;
import chat.teco.tecochat.chat.domain.chat.ChatRepository;
import chat.teco.tecochat.chat.exception.chat.ChatException;
import chat.teco.tecochat.like.chatlike.application.usecase.PushChatLikeUseCase;
import chat.teco.tecochat.like.chatlike.domain.ChatLike;
import chat.teco.tecochat.like.chatlike.domain.ChatLikeRepository;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Transactional
@Service
public class ChatLikeService implements PushChatLikeUseCase {

    private final ChatLikeRepository chatLikeRepository;
    private final ChatRepository chatRepository;

    public void pushLike(PushChatLikeCommand command) {
        Optional<Chat> optionalChat = chatRepository.findById(command.chatId());
        if (optionalChat.isEmpty()) {
            throw new ChatException(NOT_FOUND_CHAT);
        }
        Chat chat = optionalChat.get();
        chatLikeRepository.findByMemberIdAndChatId(command.memberId(), command.chatId())
                .ifPresentOrElse(
                        chatLike -> deleteLike(chat, chatLike),
                        () -> createLike(chat, command.toDomain())
                );
    }

    private void deleteLike(Chat chat, ChatLike chatLike) {
        chatLikeRepository.delete(chatLike);
        chat.decreaseLike();
    }

    private void createLike(Chat chat, ChatLike chatLike) {
        chatLikeRepository.save(chatLike);
        chat.increaseLike();
    }
}

File: src/main/kotlin/chat/teco/tecochat/application/ChatLikeService.kt
package chat.teco.tecochat.application

import chat.teco.tecochat.domain.chat.ChatRepository
import chat.teco.tecochat.domain.chatlike.ChatLike
import chat.teco.tecochat.domain.chatlike.ChatLikeRepository
import chat.teco.tecochat.query.ChatLikeQueryRepository
import chat.teco.tecochat.query.QueryChatLikeByChatIdResponse
import chat.teco.tecochat.query.QueryChatLikedByMemberIdResponse
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional


@Transactional
@Service
class ChatLikeService(
    private val chatLikeRepository: ChatLikeRepository,
    private val chatLikeQueryRepository: ChatLikeQueryRepository,
    private val chatRepository: ChatRepository,
) {

    fun pushLike(memberId: Long, chatId: Long) {
        val chat = chatRepository.getById(chatId)
        chatLikeRepository.findByMemberIdAndChatId(memberId, chatId)?.let {
            chatLikeRepository.delete(it)
            chat.decreaseLike()
            return
        }
        chatLikeRepository.save(ChatLike(memberId, chatId))
        chat.increaseLike()
    }

    fun findAllByChatId(chatId: Long): List<QueryChatLikeByChatIdResponse> =
        chatLikeQueryRepository.findAllByChatId(chatId)

    fun findAllByMemberId(chatId: Long, pageable: Pageable): Page<QueryChatLikedByMemberIdResponse> =
        chatLikeQueryRepository.findAllByMemberId(chatId, pageable)
}
